python/ray/data/tests/test_dataset_tf.py

```diff
@@ -0,0 +1,140 @@
+    )
+    def test_element_spec_dtype(self, data, expected_dtype):
+        ds = ray.data.from_items([{"spam": data, "ham": data}])
+
+        dataset = ds.to_tf(feature_columns="spam", label_columns="ham")
+
+        feature_spec, label_spec = dataset.element_spec
+        assert feature_spec.dtype == expected_dtype
+        assert label_spec.dtype == expected_dtype
+
+    def test_element_spec_shape(self):
+        ds = ray.data.from_items(8 * [{"spam": 0, "ham": 0}])
+
+        dataset = ds.to_tf(feature_columns="spam", label_columns="ham", batch_size=4)
+
+        feature_spec, label_spec = dataset.element_spec
+        assert feature_spec.shape == tf.TensorShape((None,))
+        assert label_spec.shape == tf.TensorShape((None,))
+
+        features, labels = next(iter(dataset))
+        assert features.shape == tf.TensorShape((4,))
+        assert labels.shape == tf.TensorShape((4,))
+
+    def test_element_spec_shape_with_tensors(self):
+        ds = ray.data.from_items(8 * [{"spam": np.zeros([3, 32, 32]), "ham": 0}])
+
+        dataset = ds.to_tf(feature_columns="spam", label_columns="ham", batch_size=4)
+
+        feature_spec, _ = dataset.element_spec
+        assert tuple(feature_spec.shape) == tf.TensorShape((None, 3, 32, 32))
+
+        features, labels = next(iter(dataset))
+        assert features.shape == tf.TensorShape((4, 3, 32, 32))
+        assert labels.shape == tf.TensorShape((4,))
+
+    def test_training(self):
+        def build_model() -> tf.keras.Model:
+            return tf.keras.Sequential([tf.keras.layers.Dense(1)])
+
+        def train_func():
+            strategy = tf.distribute.MultiWorkerMirroredStrategy()
+            with strategy.scope():
+                multi_worker_model = build_model()
+                multi_worker_model.compile(
+                    optimizer=tf.keras.optimizers.SGD(),
+                    loss=tf.keras.losses.mean_absolute_error,
+                    metrics=[tf.keras.metrics.mean_squared_error],
+                )
+
+            dataset = session.get_dataset_shard("train").to_tf("X", "Y", batch_size=4)
+            multi_worker_model.fit(dataset)
+
+        dataset = ray.data.from_items(8 * [{"X0": 0, "X1": 0, "Y": 0}])
+        trainer = TensorflowTrainer(
+            train_loop_per_worker=train_func,
+            preprocessor=Concatenator(exclude=["Y"], output_column_name="X"),
+            scaling_config=ScalingConfig(num_workers=2),
+            datasets={"train": dataset},
+        )
+        trainer.fit()
+
+    def test_invalid_column_raises_error(self):
+        ds = ray.data.from_items([{"spam": 0, "ham": 0}])
+        with pytest.raises(ValueError):
+            ds.to_tf(feature_columns="foo", label_columns="bar")
+
+    def test_simple_dataset_raises_error(self):
+        # `range` returns a simple dataset containing `int`s.
+        ds = ray.data.range(1)
+        with pytest.raises(NotImplementedError):
+            ds.to_tf(feature_columns="spam", label_columns="ham")
```

Comment:
Should also test that a single-tensor-column datasets (e.g. created from `read_images()` or `from_numpy()`) raises as a good error.